Sources/Load.php

```diff
@@ -851,70 +851,79 @@ function loadUserSettings()
 }
 
 /**
- * Load minimal info from members table.
+ * Load minimal user settings from members table.
  *
- * @param int $user_id The incontext user ID to get the data from.
+ * @param array $users_ids The users IDs to get the data from.
  * @return array
  * @throws Exception
  */
-function loadMinUserSettings($user_id = 0)
+function loadMinUserSettings($users_ids = [])
 {
-	global $smcFunc, $cache_enable, $modSettings, $language;
+	global $smcFunc, $modSettings, $language;
 
 	$user_settings_min = array();
 
-	if (empty($user_id) || !is_int($user_id))
+	if (empty($users_ids))
 		return $user_settings_min;
 
-	if (empty($cache_enable) || null === ($user_settings_min = cache_get_data('user_settings_min-' . $user_id, 1800)))
-	{
-		$request = $smcFunc['db_query']('', '
-				SELECT time_offset, additional_groups, id_group, id_post_group, lngfile, smiley_set, time_offset
-				FROM {db_prefix}members
-				WHERE id_member = {int:id_member}
-				LIMIT 1',
-			array(
-				'id_member' => $user_id,
-			)
-		);
+	$columns_to_load = array(
+		'id_member',
+		'time_offset',
+		'additional_groups',
+		'id_group',
+		'id_post_group',
+		'lngfile',
+		'smiley_set',
+		'time_offset'
+	);
 
-		$user_settings_min = $smcFunc['db_fetch_assoc']($request);
+	call_integration_hook('integrate_load_min_user_settings_columns', array(&$columns_to_load));
 
-		$smcFunc['db_free_result']($request);
+	$request = $smcFunc['db_query']('', '
+		SELECT {raw:columns}
```

Comment:
Abuse in what sense? if we want some flexibility via hooks then is this, a comma separated string or directly inject the implode in the query string. I prefer this method since its cleaner and passing an array offers more flexibility to mod authors..

Indeed whoever uses that hook can add anything but since the query is limited to the members table (theres no option to add other tables for example) the most they can do is break the query.

Another option is to add another type {columns} which will make some validations but theres really not much we can do to validate other than perhaps casting to string and remove empty keys.

Reply:
well... 
first you had differenty syntax across mysql and pg
secondly you could add subquery here
third build a union all select

so you can add so many tables, rows and column as you like

Reply:
Which different syntax?

Yes, thats all part of what the hook provides, thats what a hook is for to let mod authors do whatever they want.Once more, is not our duty to watch mod authors and moderate what they can't or can't do.

Reply:
The question of whether we should use `{raw:columns}` or not is a red herring. It’s no more dangerous than dynamically writing the string of SQL code directly in the source code (e.g. https://github.com/SimpleMachines/SMF2.1/blob/ec38cf4b03fc1c1e9c814437dfdbca12334e6033/Sources/MessageIndex.php#L353). 

The only way to change the content of `$columns_to_load` here is via a hooked mod, which means there’s no security issue here. An attacker would first need to gain admin privileges before he could insert malicious code into this query. But if the attacker already has admin privileges, he can already do whatever he wants anyway.

Now, it is true that a badly written mod could introduce a security hole that would allow an attacker to inject malicious code via this hook, but that’s true of every mod already and doesn’t amount to an objection to @MissAllSunday’s code here.

Reply:
different syntax: mysql guys like to use ` for columns name, alias,

Reply:
mysql works fine with or without `

Reply:
i talk about mod authors,
they don't know what they do when they use this kind.

Reply:
Please let me know when your PR for handling mysql's completely optional ` usage is done and merged to apply the changes here :)

Reply:
it's already ther: https://github.com/SimpleMachines/SMF2.1/blob/63fc00a5e9c8505bbe6a5d975251233b948f9dd0/Sources/Subs-Db-mysql.php#L278

Reply:
...and all it does is stripping them back to plain strings because SMF does not use/need them.

I mean your PR changing all instances of raw and hardcoded implodes in queries to use identifier. Theres currently one usage only.

Reply:
It allow the mod author not know the details,
on pg side it's strip them again and place " on this places:
https://github.com/SimpleMachines/SMF2.1/blob/63fc00a5e9c8505bbe6a5d975251233b948f9dd0/Sources/Subs-Db-postgresql.php#L242-L244
they are designed for this case here.

Reply:
Made a small test, did the following changes:

```php
	$columns_to_load = array(
		'id_member',
		'time_offset',
		'additional_groups',
		'`id_group`',
		'id_post_group',
		'lngfile',
		'smiley_set',
		'time_offset'
	);

	call_integration_hook('integrate_load_min_user_settings_columns', array(&$columns_to_load));

	$request = $smcFunc['db_query']('', '
		SELECT {identifier:columns}
		FROM {db_prefix}members
		WHERE id_member IN ({array_int:users_ids})',
		array(
			'users_ids' => array_map('intval', array_unique($users_ids)),
			'columns' => implode(', ', $columns_to_load)
		)
	);
```

And got a nice error:

> Database Error: Unknown column 'id_member, time_offset, additional_groups, id_group, id_post_group, lngfile, smiley_set, time_offset' in 'field list'

Your proposed change won't work when using arrays, we will have to use strings which makes the change pointless anyway because when a mod author uses the hook and sees a string with no Backticks s/he will see SMF does not use `.

Reply:
array_identifier didn't exists.
identifier does support only one column,
like all others no array_*

Reply:
Then how do you want to use identifier on this very own query?

Reply:
generate the identifier of the amount of columns,
or add array_identifier

Reply:
first one is inefficient and unnecessary the other one is out of the scope for this PR.

Reply:
When it you to mutch work,
than take the hook out.
no one is forcing you to add this,
but when you add this it's need to be robust as possible.

Reply:
Robust for who? show me exactly where and how many mod authors uses Backticks and I'll do the change.

Fact is, SMF doesn't use them, mod authors doesn't use them. This is perfectly usable as it is and its compatible with both pg and mysql. So, unless you can prove that theres a considerable amount of people who use Backticks and absolutely can't use the SMF way (and demonstrate why thy can't drop the Backticks themselves) then I'll do the change.

Reply:
tinyportal https://custom.simplemachines.org/mods/index.php?mod=97
Pretty URLs https://custom.simplemachines.org/mods/index.php?mod=636
Anti Spam https://custom.simplemachines.org/mods/index.php?mod=3851

3 of top 5 addons use backticks,
so you see the mod author build ther query based on phpmyadmin,
ther is the default to use backticks so they learn, "backticks are good".

Reply:
Problems in poorly written mods need to be fixed in the mods, not in this PR. 

Alternatively, @albertlast, if you want SMF to automatically remove those backticks from SQL statements, please feel free to add something to Subs-Db-Postgresql.php to do so.

Reply:
@Sesquipedalian the identifier do this already,
you only had to use them.

Reply:
Did a quick search for backticks on TinyPortal 1.6.5. The only place where they use backticks is on their install.php file.

What I want you to understand is this, the hook priimary usage its for adding column names, column names can be easily and uniformly represented by simple string keys, this is why this:

	$columns_to_load = array(
		'id_member',
		'time_offset',
		'additional_groups',
		'id_group',
		'id_post_group',
		'lngfile',
		'smiley_set',
		'time_offset'
	);
```
Is a perfectly valid way to add column names. Remember, the primary usage for this hook is to add column names. Now, if a mod author wants to include a subquery or other SQL syntax  then they will have to write their query **accordingly** to how SMF builds queries. This makes sense because whenever you want to integrate with something you adapt to their way, **no the other way around**

It is far more easy for mod authors to adapt to a single unified way for writing queries than add support for every sql engine we support.

And I already spent way to much time on this, not gonna comment any further.